lib/jsdom/living/nodes/HTMLInputElement-impl.js

```diff
@@ -590,6 +632,209 @@ class HTMLInputElementImpl extends HTMLElementImpl {
+        valueMissing: () => {
+          if (!this.hasAttribute("required")) {
+            return false;
+          }
+          if (this.type === "checkbox") {
+            // https://html.spec.whatwg.org/multipage/input.html#checkbox-state-(type=checkbox)
+            // Constraint validation: If the element is required and its checkedness is
+            // false, then the element is suffering from being missing.
+            return !this.checked;
+          } else if (this.type === "radio") {
+            // https://html.spec.whatwg.org/multipage/input.html#radio-button-state-(type=radio)
+            // Constraint validation: If an element in the radio button group is required,
+            // and all of the input elements in the radio button group have a checkedness
+            // that is false, then the element is suffering from being missing.
+            return !this._isRadioGroupChecked();
+          }
+          return this.value === "";
+        },
+
+        // https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#suffering-from-being-too-long
+        // https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#attr-fe-maxlength
+        // jsdom has no way at the moment to emulate a user interaction, so tooLong/tooShort have
+        // to be set to false.
+        tooLong: () => false,
+
+        // https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#suffering-from-being-too-short
+        // https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#attr-fe-minlength
+        tooShort: () => false,
+
+        // https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#suffering-from-an-overflow
+        // https://html.spec.whatwg.org/multipage/input.html#attr-input-max
+        rangeOverflow: () => this._attributeApplies("max") && this._maximum !== null &&
+          this._parsedValue > this._maximum,
+
+        // https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#suffering-from-an-underflow
+        // https://html.spec.whatwg.org/multipage/input.html#attr-input-min
+        rangeUnderflow: () => this._attributeApplies("min") && this._minimum !== null &&
+          this._parsedValue < this._minimum,
+
+        // https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#suffering-from-a-pattern-mismatch
+        patternMismatch: () => {
+          if (!this.hasAttribute("pattern") || !this._attributeApplies("pattern") || this.value === "") {
+            return false;
+          }
+          let regexp;
+          try {
+            regexp = new RegExp(this.getAttribute("pattern"), "u");
+          } catch (e) {
+            return false;
+          }
+          const multiple = this._attributeApplies("multiple") && this.hasAttribute("multiple");
+          return !regexpMatches({ regexp, testValue: this.value, split: multiple });
```

Comment:
So, you have applied the concept of an elements "values" too widely here. Not every input element can have "values". Only email and file inputs can. So you shouldn't always split them. I think this should be special-cased for emails, not use a generic regexpMatches helper, as I noted previously.

Reply:
I was under the impression that both `type="file"` and `type="email"` support the multiple attribute and hence "values". Due to that there's `this._attributeApplies("multiple")` above which resolves to:
```
const applicableTypesForAttribute = {
...
  multiple: new Set(["email", "file"]),
...
};
```

Reply:
updated to only do the split for emails